crates/e2e/src/setup/solver.rs

```diff
@@ -0,0 +1,245 @@
+    observe::tracing::initialize_reentrant(&config.log);
+    tracing::info!("running mock solver engine with {config:#?}");
+
+    let solver = Mock::new(config.solution);
+
+    Api {
+        addr: config.addr,
+        solver,
+    }
+    .serve(bind, shutdown_signal())
+    .await
+    .unwrap();
+}
+
+pub async fn solve(
+    state: axum::extract::State<Arc<Mock>>,
+    Json(auction): Json<Auction>,
+) -> (axum::http::StatusCode, Json<Response<Solutions>>) {
+    let handle_request = async {
+        let auction_id = auction.id.unwrap_or_default();
+        let solutions = state
+            .solve(auction)
+            .instrument(tracing::info_span!("auction", id = %auction_id))
+            .await;
+
+        tracing::trace!(?auction_id, ?solutions);
+
+        (axum::http::StatusCode::OK, Json(Response::Ok(solutions)))
+    };
+
+    handle_request
+        .instrument(tracing::info_span!("/solve"))
+        .await
+}
+
+#[derive(Debug, Serialize)]
+#[serde(untagged)]
+pub enum Response<T> {
+    Ok(T),
+    Err(Error),
+}
+
+#[derive(Debug, Serialize)]
+pub struct Error {
+    pub message: &'static str,
+}
+
+impl From<&'static str> for Error {
+    fn from(message: &'static str) -> Self {
+        Self { message }
+    }
+}
+
+const REQUEST_BODY_LIMIT: usize = 10 * 1024 * 1024;
+
+pub struct Api {
+    pub addr: SocketAddr,
+    pub solver: Mock,
+}
+
+impl Api {
+    pub async fn serve(
+        self,
+        bind: Option<oneshot::Sender<SocketAddr>>,
+        shutdown: impl Future<Output = ()> + Send + 'static,
+    ) -> Result<(), hyper::Error> {
+        let app = axum::Router::new()
+            .layer(tower::ServiceBuilder::new().layer(
+                tower_http::limit::RequestBodyLimitLayer::new(REQUEST_BODY_LIMIT),
+            ))
+            .route("/solve", axum::routing::post(solve))
+            .layer(
+                tower::ServiceBuilder::new().layer(tower_http::trace::TraceLayer::new_for_http()),
+            )
+            .with_state(Arc::new(self.solver))
+            // axum's default body limit needs to be disabled to not have the default limit on top of our custom limit
+            .layer(axum::extract::DefaultBodyLimit::disable());
+
+        let make_svc = observe::make_service_with_task_local_storage!(app);
+
+        let server = axum::Server::bind(&self.addr).serve(make_svc);
+        if let Some(bind) = bind {
+            let _ = bind.send(server.local_addr());
+        }
+
+        server.with_graceful_shutdown(shutdown).await
+    }
+}
+
+#[cfg(unix)]
+async fn shutdown_signal() {
+    // Intercept main signals for graceful shutdown.
+    // Kubernetes sends sigterm, whereas locally sigint (ctrl-c) is most common.
+    let mut interrupt = unix::signal(SignalKind::interrupt()).unwrap();
+    let mut terminate = unix::signal(SignalKind::terminate()).unwrap();
+    tokio::select! {
+        _ = interrupt.recv() => (),
+        _ = terminate.recv() => (),
+    };
+}
+
+#[derive(Clone, Debug)]
```

Comment:
Eventually this should be moved into a separate function.
We could have 1 file implementing the mock solver and another file that contains a bunch of helper functions to build the solution that the solver is supposed to return (like this one).

Reply:
Yes, I fully agree, once we have all the tests I want to clean it up for merge :ok_hand: